pkg/reconciler/tenancy/initialization/apibinder_initializer_controller.go

```diff
@@ -0,0 +1,304 @@
+		}
+
+		parentClusterName, clusterWorkspaceName := clusters.SplitClusterAwareKey(clusterName.String())
+		logger.Error(err, "failed to get ClusterWorkspace from lister", "parentCluster", parentClusterName, "clusterWorkspace", clusterWorkspaceName)
+		return // nothing we can do here
+	}
+
+	b.enqueueClusterWorkspace(clusterWorkspace, logger, fmt.Sprintf(" because of APIBinding %s", metaObj.GetName()))
+}
+
+// enqueueClusterWorkspaceType enqueues all clusterworkspaces (which are only those that are initializing, because of
+// how the informer is supposed to be configured) whenever a clusterworkspacetype changes. If a clusterworkspacetype
+// had a typo in the default set of apibindings, there is a chance the requeuing here would pick up a fix.
+func (b *APIBinder) enqueueClusterWorkspaceType(obj interface{}, logger logr.Logger) {
+	cwt, ok := obj.(*tenancyv1alpha1.ClusterWorkspaceType)
+	if !ok {
+		runtime.HandleError(fmt.Errorf("obj is supposed to be a ClusterWorkspaceType, but is %T", obj))
+		return
+	}
+
+	if len(cwt.Spec.DefaultAPIBindings) == 0 {
+		return
+	}
+
+	list, err := b.listClusterWorkspaces()
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("error listing clusterworkspaces: %w", err))
+	}
+
+	for _, ws := range list {
+		b.enqueueClusterWorkspace(ws, logger, fmt.Sprintf(" because of ClusterWorkspaceType %s|%s", logicalcluster.From(cwt), cwt.GetName()))
+	}
+}
+
+func (b *APIBinder) startWorker(ctx context.Context) {
+	for b.processNextWorkItem(ctx) {
+	}
+}
+
+func (b *APIBinder) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer b.queue.ShutDown()
+	logger := logging.WithReconciler(klog.FromContext(ctx), ControllerName)
+	ctx = klog.NewContext(ctx, logger)
+
+	logger.Info("Starting controller")
+	defer logger.Info("Shutting down controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(ctx, b.startWorker, time.Second)
+	}
+	<-ctx.Done()
+}
+
+func (b *APIBinder) ShutDown() {
+	b.queue.ShutDown()
+}
+
+func (b *APIBinder) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := b.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	logger := logging.WithQueueKey(klog.FromContext(ctx), key)
+	ctx = klog.NewContext(ctx, logger)
+	logger.V(1).Info("processing key")
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer b.queue.Done(key)
+
+	if err := b.process(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("%s: failed to sync %q, err: %w", ControllerName, key, err))
+		b.queue.AddRateLimited(key)
+		return true
+	}
+
+	b.queue.Forget(key)
+	return true
+}
+
+func (b *APIBinder) process(ctx context.Context, key string) error {
+	logger := klog.FromContext(ctx)
+	startTime := time.Now()
+
+	defer func() {
+		logger.V(4).Info("finished syncing", "duration", time.Since(startTime))
+	}()
+
+	parent, workspace := clusters.SplitClusterAwareKey(key)
```

Comment:
You created the key with `key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)` - please decode it appropriately, instead of assuming that we can apply SplitClusterAwareKey to it directly